components/de_amp/browser/de_amp_throttle.cc

```diff
@@ -0,0 +1,71 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/de_amp/browser/de_amp_throttle.h"
+
+#include <utility>
+
+#include "base/feature_list.h"
+#include "brave/components/de_amp/browser/de_amp_service.h"
+#include "brave/components/de_amp/browser/de_amp_url_loader.h"
+#include "brave/components/de_amp/common/features.h"
+#include "components/content_settings/core/browser/host_content_settings_map.h"
+#include "components/content_settings/core/common/content_settings.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/pending_receiver.h"
+#include "mojo/public/cpp/bindings/pending_remote.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+
+namespace de_amp {
+
+// static
+std::unique_ptr<DeAmpThrottle> DeAmpThrottle::MaybeCreateThrottleFor(
+    scoped_refptr<base::SingleThreadTaskRunner> task_runner,
+    DeAmpService* service,
+    content::WebContents* contents) {
+  if (!service->IsEnabled()) {
+    return nullptr;
+  }
+  return std::make_unique<DeAmpThrottle>(task_runner, service, contents);
+}
+
+DeAmpThrottle::DeAmpThrottle(
+    scoped_refptr<base::SingleThreadTaskRunner> task_runner,
+    DeAmpService* service,
+    content::WebContents* contents)
+    : task_runner_(std::move(task_runner)),
+      service_(service),
+      contents_(contents) {}
+
+DeAmpThrottle::~DeAmpThrottle() = default;
+
+void DeAmpThrottle::WillProcessResponse(
+    const GURL& response_url,
+    network::mojom::URLResponseHead* response_head,
+    bool* defer) {
+  VLOG(2) << "deamp throttling: " << response_url;
+  *defer = true;
```

Comment:
Can we limit this to some partucular set of urls?

Reply:
Do you mean limit the de-amp to a set of URLs? Unfortunately not, there's no way to tell from the URL if the page is an AMP page. Maintaining a hardcoded list of URLs that are approved for de-amping would be hard.

Reply:
I hope this will not create a performance impact.
Do we really need the whole 64Kb?

Reply:
This is definitely a heuristic - in my testing I came across at least one website that had a <head> larger than 32kb. I can add a TODO to come up with a more rigorous number.

Reply:
@bridiver you told that we need only the first chunk..

Reply:
We should still only operate on the first chunk, just that the first chunk will be 64kb.

Reply:
@atuchin-m we read whatever is available up to 64K, but we don't wait for 64K to make a decision, we just wait for whatever is available on the first read call. After the first read call we either find amp and redirect or give up and pass through

Reply:
`ReadData` reads up to the limit and updates it with the value that was actually read

Reply:
@bridiver 
If so, why we and chromium use `size_t start_size = buffered_body_.size();` not `size_t start_size = 0`?

Reply:
@atuchin-m because the read and check happen in two different methods. It's a bit confusing because of the way it's broken up (which was done to share code with speedreader), but I went through this last night with @ShivanKaul to verify that it was all correct.